cmd/snap-bootstrap/cmd_initramfs_mounts.go

```diff
@@ -319,115 +473,96 @@ func selectPartitionMatchingKernelDisk(dir, fallbacklabel string) error {
-		}
-		if !isKernelMounted {
-			whichTypes = append(whichTypes, snap.TypeKernel)
-		}
-		if !isSnapdMounted {
-			whichTypes = append(whichTypes, snap.TypeSnapd)
+
+	// TODO:UC20: do we need more cross checks here?
+	for _, essentialSnap := range essSnaps {
+		dir := snapTypeToMountDir[essentialSnap.EssentialType]
+		// TODO:UC20: we need to cross-check the kernel path with snapd_recovery_kernel used by grub
+		opts := &SystemdMountOptions{
+			// always persist the snaps mount
+			SurvivesPivotRoot: true,
 		}
-		essSnaps, err := mst.RecoverySystemEssentialSnaps("", whichTypes)
+		err := doSystemdMount(essentialSnap.Path, filepath.Join(boot.InitramfsRunMntDir, dir), opts)
 		if err != nil {
-			return false, fmt.Errorf("cannot load metadata and verify essential bootstrap snaps %v: %v", whichTypes, err)
-		}
-
-		// TODO:UC20: do we need more cross checks here?
-		for _, essentialSnap := range essSnaps {
-			switch essentialSnap.EssentialType {
-			case snap.TypeBase:
-				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(boot.InitramfsRunMntDir, "base"))
-			case snap.TypeKernel:
-				// TODO:UC20: we need to cross-check the kernel path with snapd_recovery_kernel used by grub
-				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(boot.InitramfsRunMntDir, "kernel"))
-			case snap.TypeSnapd:
-				fmt.Fprintf(stdout, "%s %s\n", essentialSnap.Path, filepath.Join(boot.InitramfsRunMntDir, "snapd"))
-			}
+			return err
 		}
 	}
 
+	// TODO:UC20: after we have the kernel and base snaps mounted, we should do
+	//            the bind mounts from the kernel modules on top of the base
+	//            mount and delete the corresponding systemd units from the
+	//            initramfs layout
+
 	// 2.3. mount "ubuntu-data" on a tmpfs
-	isMounted, err = mst.IsMounted(boot.InitramfsDataDir)
-	if err != nil {
-		return false, err
-	}
-	if !isMounted {
-		fmt.Fprintf(stdout, "--type=tmpfs tmpfs %s\n", boot.InitramfsDataDir)
-		return false, nil
+	opts := &SystemdMountOptions{
+		// always persist the tmpfs ubuntu-data mount
+		SurvivesPivotRoot: true,
+		IsTmpfs:           true,
 	}
-
-	return true, nil
+	return doSystemdMount("tmpfs", boot.InitramfsDataDir, opts)
 }
 
 func generateMountsModeRun(mst *initramfsMountsState) error {
 	// 1. mount ubuntu-boot
-	isMounted, err := mst.IsMounted(boot.InitramfsUbuntuBootDir)
+	err := selectPartitionMatchingKernelDisk(boot.InitramfsUbuntuBootDir, "ubuntu-boot")
 	if err != nil {
 		return err
 	}
-	if !isMounted {
-		return selectPartitionMatchingKernelDisk(boot.InitramfsUbuntuBootDir, "ubuntu-boot")
-	}
 
 	// 2. mount ubuntu-seed
-	isMounted, err = mst.IsMounted(boot.InitramfsUbuntuSeedDir)
+	// TODO:UC20: use the ubuntu-boot partition as a reference for what
+	//            partition to mount for ubuntu-seed
+	opts := &SystemdMountOptions{
+		// TODO: do we need to always run fsck on ubuntu-seed in run mode?
+		//       would it be safer to not touch ubuntu-seed at all for boot?
```

Comment:
yes, maybe it would be best to leave that only to recover mode??

Reply:
sure, I will make it so that in recover mode, we run fsck on ubuntu-seed, all other situations we don't run fsck on ubuntu-seed.

Reply:
actually, thinking about it, @xnox had opinions on this when we discussed it but I don't remember what they were

@xnox do you have an opinion on when (and on what partitions) should we run fsck ?

Reply:
I implemented this such that we don't fsck ubuntu-seed in run mode specifically, happy to change this though if folks feel differently

Reply:
let' see what @xnox think,  we should also remember that seed is a FAT partition

Reply:
In run-mode we mount ubuntu-seed filesystem read-write, it means we must run fsck on it. Otherwise it may be mounted read-only, and then chooser will not work.

Reply:
@xnox sorry I don't follow, we are mounting it rw, how could it accidentally get mounted ro a la "Otherwise it may be mounted read-only" ?